src/zarr/group.py

```diff
@@ -269,26 +287,68 @@ async def update_attributes(self, new_attributes: Dict[str, Any]):
     def __repr__(self):
         return f"<AsyncGroup {self.store_path}>"
 
-    async def nchildren(self) -> int:
-        raise NotImplementedError
-
-    async def children(self) -> AsyncIterator[Union[AsyncArray, AsyncGroup]]:
-        raise NotImplementedError
-
-    async def contains(self, child: str) -> bool:
+    async def nmembers(self) -> int:
         raise NotImplementedError
 
-    async def group_keys(self) -> AsyncIterator[str]:
-        raise NotImplementedError
+    async def members(self) -> AsyncGenerator[tuple[str, AsyncArray | AsyncGroup], None]:
+        """
+        Returns an AsyncGenerator over the arrays and groups contained in this group.
+        This method requires that `store_path.store` supports directory listing.
+
+        The results are not guaranteed to be ordered.
+        """
+        if not self.store_path.store.supports_listing:
+            msg = (
+                f"The store associated with this group ({type(self.store_path.store)}) "
+                "does not support listing, "
+                "specifically via the `list_dir` method. "
+                "This function requires a store that supports listing."
+            )
 
-    async def groups(self) -> AsyncIterator[AsyncGroup]:
-        raise NotImplementedError
+            raise ValueError(msg)
+        subkeys = await self.store_path.store.list_dir(self.store_path.path)
+        # would be nice to make these special keys accessible programmatically,
+        # and scoped to specific zarr versions
+        subkeys_filtered = filter(lambda v: v not in ("zarr.json", ".zgroup", ".zattrs"), subkeys)
+        # is there a better way to schedule this?
+        for subkey in subkeys_filtered:
+            try:
+                yield (subkey, await self.getitem(subkey))
+            except KeyError:
+                # keyerror is raised when `subkey` names an object (in the object storage sense),
+                # as opposed to a prefix, in the store under the prefix associated with this group
+                # in which case `subkey` cannot be the name of a sub-array or sub-group.
+                logger.warning(
+                    "Object at %s is not recognized as a component of a Zarr hierarchy.", subkey
+                )
+                pass
 
-    async def array_keys(self) -> AsyncIterator[str]:
+    async def contains(self, member: str) -> bool:
         raise NotImplementedError
 
+    # todo: decide if this method should be separate from `groups`
+    async def group_keys(self) -> AsyncGenerator[str, None]:
+        async for key, value in self.members():
+            if isinstance(value, AsyncGroup):
+                yield key
```

Comment:
A **future** optimization here would be to avoid constructing the `AsyncGroup` and instead just generate the key based on filtering the metadata doc. As it is currently written, we construct all the members as `AsyncGroup | AsyncArray`, only to return their name.

Reply:
Additionally, we could allow `Group.members` to filter to `only='groups'` so that we never need construct arrays in these cases.

Reply:
yes, i think checking the metadata should be much cheaper than instantiating `AsyncArray` / `AsyncGroup`. I will have to think about what those changes would mean for the `AsyncGroup.members` API, since that's what `group_keys` relies on.